File: example/env/flutter/material/materialApp.ts
import { Widget } from './../widget';
import { StatelessWidget } from './../widgets/statelessWidget';

interface MaterialAppProps {
    initialRoute: string;
    home: Widget;
}

declare const flutter: {
    material: {
        materialApp: (this: void, props: MaterialAppProps) => { tag: string };
    }
}

export class MaterialApp extends StatelessWidget {
    public props: MaterialAppProps;
    public constructor(props: MaterialAppProps) {
        super();
        this.props = props;
    }

    public build(): Widget {
        return flutter.material.materialApp(this.props);
    }
}

File: lib/cfr/builtins/flutter/material/materialApp.dart
import 'package:hydro_sdk/hydroState.dart';
import 'package:hydro_sdk/cfr/vm/context.dart';
import 'package:hydro_sdk/cfr/builtins/flutter/syntheticBox.dart';
import 'package:hydro_sdk/cfr/vm/table.dart';
import 'package:flutter/material.dart';

loadMaterialApp({@required HydroState luaState, @required HydroTable table}) {
  table["materialApp"] = makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      MaterialApp(
        initialRoute: args[0]["initialRoute"],
        home: maybeUnwrapAndBuildArgument<Widget>(args[0]["home"],
            parentState: luaState),
        title: args[0]["title"],
      )
    ];
  });
}
